zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/AdHocSubProcessProcessor.java

```diff
@@ -203,26 +204,66 @@ private void activateElements(
                 stateTransitionBehavior.activateChildInstance(context, elementToActivate));
   }
 
+  @Override
+  public Either<Failure, ?> beforeExecutionPathCompleted(
+      final ExecutableAdHocSubProcess adHocSubProcess,
+      final BpmnElementContext adHocSubProcessContext,
+      final BpmnElementContext childContext) {
+    final Expression completionConditionExpression = adHocSubProcess.getCompletionCondition();
+    if (completionConditionExpression == null) {
+      return Either.right(null);
+    }
+
+    return expressionProcessor
+        .evaluateBooleanExpression(
+            completionConditionExpression, adHocSubProcessContext.getElementInstanceKey())
+        .mapLeft(
+            failure ->
+                new Failure(
+                    "Failed to evaluate completion condition. " + failure.getMessage(),
+                    ErrorType.EXTRACT_VALUE_ERROR));
+  }
+
   @Override
   public void afterExecutionPathCompleted(
-      final ExecutableAdHocSubProcess element,
+      final ExecutableAdHocSubProcess adHocSubProcess,
       final BpmnElementContext adHocSubProcessContext,
       final BpmnElementContext childContext,
       final Boolean satisfiesCompletionCondition) {
-
-    if (stateBehavior.canBeCompleted(childContext)) {
-      stateTransitionBehavior.completeElement(adHocSubProcessContext);
+    if (satisfiesCompletionCondition == null) {
+      if (stateBehavior.canBeCompleted(childContext)) {
+        stateTransitionBehavior.completeElement(adHocSubProcessContext);
+      }
+    } else {
+      if (satisfiesCompletionCondition) {
+        if (adHocSubProcess.isCancelRemainingInstancesEnabled()) {
+          final boolean hasNoActiveChildren =
+              stateTransitionBehavior.terminateChildInstances(adHocSubProcessContext);
+          if (hasNoActiveChildren) {
+            stateTransitionBehavior.completeElement(adHocSubProcessContext);
+          }
+        } else {
+          if (stateBehavior.canBeCompleted(childContext)) {
+            stateTransitionBehavior.completeElement(adHocSubProcessContext);
+          }
+        }
+      }
     }
   }
 
   @Override
   public void onChildTerminated(
-      final ExecutableAdHocSubProcess element,
+      final ExecutableAdHocSubProcess adHocSubProcess,
       final BpmnElementContext adHocSubProcessContext,
       final BpmnElementContext childContext) {
-
-    if (stateBehavior.canBeTerminated(childContext)) {
-      terminate(element, adHocSubProcessContext);
+    if (adHocSubProcessContext.getIntent() == ProcessInstanceIntent.ELEMENT_TERMINATING) {
```

Comment:
This should be the same logic as handled in [MultiInstanceBodyProcessor](https://github.com/camunda/camunda/blob/main/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java)